src/neuroglancer/ui/segment_select_tools.ts

```diff
@@ -0,0 +1,142 @@
+/**
+ * /**
+ * @license
+ * This work is a derivative of the Google Neuroglancer project,
+ * Copyright 2016 Google Inc.
+ * The Derivative Work is covered by
+ * Copyright 2021 Howard Hughes Medical Institute
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {SegmentationUserLayer} from 'neuroglancer/segmentation_user_layer';
+import {removeChildren} from 'neuroglancer/util/dom';
+import {makeToolActivationStatusMessageWithHeader, registerLayerTool, Tool, ToolActivation} from 'neuroglancer/ui/tool';
+import { EventActionMap } from '../util/event_action_map';
+
+export const SELECT_SEGMENTS_TOOLS_ID = 'selectSegments';
+
+const selectEvent = 'mousedown0';
+const deselectEvent = 'mousedown2';
+const SELECT_SEGMENTS_INPUT_EVENT_MAP = EventActionMap.fromObject({
+  [`at:shift?+${selectEvent}`]: 'drag-select-segments',
+  [`at:shift?+${deselectEvent}`]: 'drag-deselect-segments',
+  'at:shift?+mouseup0': 'deactivate-drag-select-segments',
```

Comment:
Rather than use separate mouseup events to end the drag, which can be fragile (e.g. mouseup may happen outside the browser window), you can use `startRelativeMouseDrag` (from util/mouse_drag.ts) which installs its own mouseup handlers to end the drag, as is used elsewhere in neuroglancer for mouse drag handling.

`startRelativeMouseDrag` will also invoke a callback on each mouse move event, which in this case you can ignore.

Reply:
Before pushing the commit, I'd like to make sure I'm using startRelativeMouseDrag correctly:

```
  const startSelecting = (event: ActionEvent<MouseEvent>) => {
      updateStatus();
      trySelectSegment();
      activation.registerDisposer(
        layer.displayState.segmentSelectionState.changed.add(trySelectSegment));
      activation.registerDisposer(
        layer.displayState.segmentationGroupState.value.visibleSegments.changed.add(updateStatus));
      startRelativeMouseDrag(
        event.detail,
        (_event, _deltaX, _deltaY) => {},
        (_event) => {
          state = ToolState.IGNORE;
          updateStatus();
        }
      );
    };
```

Since startRelativeMouseDrag requires mouse event, I need to cast the event in action binding like this:
```
  activation.bindAction('drag-select-segments', event => {
      event.stopPropagation();
      state = ToolState.SELECT;
      startSelecting(event as ActionEvent<MouseEvent>);
    });
```
Hope this will not cause any problem.

Reply:
Yes, that's right --- rather than using `as` you can also do:

`activation.bindAction('drag-select-segments', (event: ActionEvent<MouseEvent>) => { ... });`

But either way it is true that if you were to bind something that is not a mouse event to trigger `drag-select-segments` then it would not work correctly.